### PR TITLE
Make plugin-specific env take precedence over sys env

### PR DESCRIPTION
**Pull request by tomhjp**
Unless `SkipHostEnv` is set, go-plugin layers `os.Environ()` over the top of any environment provided in the `ClientConfig`'s command: https://github.com/hashicorp/go-plugin/blob/586d14f3dcef1eb42bfb7da4c7af102ec6638668/client.go#L657

While it's too big of a breaking change to switch that behaviour in go-plugin, for Vault it seems pretty obvious that users would expect the env they specify for plugins to take precedence over the system env. For example, if most of Vault should use one proxy via HTTP_PROXY, but one plugin needs to use a different proxy, it should be possible to override the base setting when registering the plugin.

I've added a flag to opt out of the behaviour change though, and also implemented log warnings to help any users concerned about unexpected changes in behaviour.

### CODE REVIEW

File: sdk/helper/pluginutil/run_config.go [hunk 1]
```diff
@@ -107,7 +107,35 @@ func (rc runConfig) generateCmd(ctx context.Context) (cmd *exec.Cmd, clientTLSCo
...
+	}
+
+	if rc.image == "" {
+		// go-plugin has always overridden user-provided env vars with the OS
+		// (Vault process) env vars, but we want plugins to be able to override
+		// the Vault process env. We don't want to make a breaking change in
+		// go-plugin so always set SkipHostEnv and replicate the legacy behavior
+		// ourselves if user opts in.
+		if legacy, _ := strconv.ParseBool(os.Getenv(PluginUseLegacyEnvLayering)); legacy {
+			// Env vars are layered as follows, with later entries overriding
```
Comment (austingebauer) [inline sdk/helper/pluginutil/run_config.go:120]: 👍

File: vault/plugincatalog/plugin_catalog.go [hunk 2]
```diff
@@ -169,6 +171,31 @@ func SetupPluginCatalog(ctx context.Context, in *PluginCatalogInput) (*PluginCat
...
+			for k := range pluginKeys {
+				if _, ok := osKeys[k]; ok {
+					conflicts = true
+					logger.Warn("conflict between system and plugin environment variable", "type", plugin.Type, "name", plugin.Name, "version", plugin.Version, "variable", k)
+				}
+			}
+		}
+		if conflicts {
+			logger.Warn("conflicts found between system and plugin environment variables, "+
+				"system environment variables will take precendence until flag is disabled",
```
Comment (austingebauer) [inline vault/plugincatalog/plugin_catalog.go:192]:
```suggestion
				"system environment variables will take precedence until flag is disabled",
```

File: vault/plugincatalog/plugin_catalog_test.go [hunk 3]
```diff
@@ -69,6 +70,115 @@ func testPluginCatalog(t *testing.T) *PluginCatalog {
 	return pluginCatalog
 }
 
+type warningCountLogger struct {
```
Comment (austingebauer) [inline vault/plugincatalog/plugin_catalog_test.go:73]: Neat pattern to test on the logs!

**Review by austingebauer — APPROVED** [review]
Change seems reasonable to me 👍

**Comment by tomhjp** [comment]
Thanks!

**Pull request merged**